pyhf/pdf.py

```diff
@@ -131,6 +137,173 @@ def __init__(self, spec, **config_kwargs):
+                    if mtype == 'histosys':
+                        lo_data = thismod['data']['lo_data'] if thismod else nom
+                        hi_data = thismod['data']['hi_data'] if thismod else nom
+                        maskval = True if thismod else False
+                        mega_mods[s][m]['data']['lo_data'] += lo_data
+                        mega_mods[s][m]['data']['hi_data'] += hi_data
+                        mega_mods[s][m]['data']['mask']    += [maskval]*len(nom) #broadcasting
+                        pass
+                    elif mtype == 'normsys':
+                        maskval = True if thismod else False
+                        lo_factor = thismod['data']['lo'] if thismod else 1.0
+                        hi_factor = thismod['data']['hi'] if thismod else 1.0
+                        mega_mods[s][m]['data']['lo']   += [lo_factor]*len(nom) #broadcasting
+                        mega_mods[s][m]['data']['hi']   += [hi_factor]*len(nom)
+                        mega_mods[s][m]['data']['mask'] += [maskval]  *len(nom) #broadcasting
+                    elif mtype == 'normfactor':
+                        maskval = True if thismod else False
+                        mega_mods[s][m]['data']['mask'] += [maskval]*len(nom) #broadcasting
+                    elif mtype == 'staterror':
+                        uncrt = thismod['data'] if thismod else [0.0]*len(nom)
+                        maskval = [True if thismod else False]*len(nom)
+                        mega_mods[s][m]['data']['mask']  += maskval
+                        mega_mods[s][m]['data']['uncrt'] += uncrt
+                    elif mtype == 'shapefactor':
+                        maskval = True if thismod else False
+                        mega_mods[s][m]['data']['mask'] += [maskval]*len(nom) #broadcasting
+                    else:
+                        raise RuntimeError
+            sample_dict = {
+                'name': 'mega_{}'.format(s),
+                'nom': mega_nom,
+                'modifiers': list(mega_mods[s].values())
+            }
+            mega_samples[s] = sample_dict
+        self.mega_samples = mega_samples
+        self.mega_mods    = mega_mods
+
+    def prep(self):
+        import numpy as np
+        self.normsys_histoset = np.asarray([
+            [
+                [
+                    self.mega_mods[s][m]['data']['lo'],
+                    [1.]*len(self.mega_samples[s]['nom']),
+                    self.mega_mods[s][m]['data']['hi'],
+                ]
+                for s in self.do_samples
+            ] for m,mtype in self.do_mods if mtype == 'normsys' 
+        ])
+        self.normsys_mask = np.asarray([
+            [
+                [
+                    self.mega_mods[s][m]['data']['mask'],
+                ]
+                for s in self.do_samples
+            ] for m,mtype in self.do_mods if mtype == 'normsys' 
+        ])
+        self.normsys_default = np.ones(self.normsys_mask.shape)
+
```

Comment:
here we create a mask for the affected bins and prep the default values that should be applied to those not affeced